include/matx/core/iterator.h

```diff
@@ -0,0 +1,230 @@
+  using scalar_type = typename OperatorType::scalar_type;
+  // using stride_type = std::conditional_t<is_tensor_view_v<OperatorType>, typename OperatorType::desc_type::stride_type,
+  //                         index_t>;
+  using stride_type = index_t;
+  using pointer = value_type*;
+  using reference = value_type&;
+  using iterator_category = std::random_access_iterator_tag;
+  using difference_type = typename std::iterator<iterator_category, value_type>::difference_type;
+
+  __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ RandomOperatorIterator(const OperatorType &t) : t_(t), offset_(0) { }
+  __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ RandomOperatorIterator(const OperatorType &t, stride_type offset) : t_(t), offset_(offset) {}
+
+  /**
+   * @brief Dereference value at a pre-computed offset
+   * 
+   * @return Value at offset 
+   */
+  [[nodiscard]] __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ value_type operator*() const
+  {
+    if constexpr (OperatorType::Rank() == 0) {
+      return t_.operator()();
+    }
+    else {
+      auto arrs = detail::GetIdxFromAbs(t_, offset_);
+      return detail::mapply([&](auto &&...args) {
+          return t_.operator()(args...);
+        }, arrs);     
+    }
+  }  
+
+
+  [[nodiscard]] __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ self_type operator+(difference_type offset) const
+  {
+    return self_type{t_, offset_ + offset};
+  }
+
+  [[nodiscard]] __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ value_type operator[](difference_type offset) const
+  {
+    return *self_type{t_, offset_ + offset};
+  }  
+
+  __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__  self_type operator++(int)
+  {
+      self_type retval = *this;
+      offset_++;
+      return retval;
+  }  
+
+  __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ self_type operator++()
+  {
+      offset_++;
+      return *this;
+  }  
+
+  __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ self_type& operator+=(difference_type offset)
+  {
+      offset_ += offset;
+      return *this;
+  }
+
+  __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ self_type operator-(difference_type offset) const
+  {
+      return self_type{t_, offset_ - offset};
+  }
+
+
+  __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ self_type& operator-=(difference_type offset)
+  {
+      offset_ -= offset;
+      return *this;
+  }  
+
+  static __MATX_INLINE__ constexpr __MATX_HOST__ __MATX_DEVICE__ int32_t Rank() {
+    return OperatorType::Rank();
+  }
+
+  constexpr __MATX_INLINE__ __MATX_HOST__ __MATX_DEVICE__ index_t Size(int dim) const
+  {
+    return t_.Size(dim);
+  }  
+
+  typename detail::base_type_t<OperatorType> t_;
+  stride_type offset_;  
+};
+
+
+
+/**
+ * @brief Iterator around operators for libraries that can take iterators as output (CUB).
+ * 
+ * @tparam T Data type
+ * @tparam RANK Rank of tensor
+ * @tparam Desc Descriptor for tensor
+ * 
+ */
+template <typename OperatorType>
+struct RandomOperatorOutputIterator {
```

Comment:
Do we need both input and output operator or can we have one with both lvalue and rvalue semantics?

Reply:
I tried getting rid of the output version and got a ton of compiler warnings. I'll leave it as TBD for now